core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala

```diff
@@ -91,17 +103,25 @@ private[spark] class HeartbeatReceiver(sc: SparkContext, scheduler: TaskSchedule
         scheduler.executorLost(executorId, SlaveLost("Executor heartbeat " +
           s"timed out after ${now - lastSeenMs} ms"))
         if (sc.supportDynamicAllocation) {
-          sc.killExecutor(executorId)
+          // Asynchronously kill the executor to avoid blocking the current thread
```

Comment:
No.

Here is the execution of `killExecutor`:

SparkContext.killExecutor -> SparkContext.killExecutors -> CoarseGrainedSchedulerBackend.killExecutors -> CoarseGrainedSchedulerBackend.doKillExecutors

If CoarseGrainedSchedulerBackend is YarnSchedulerBackend, it will `askWithReply` KillExecutors to YarnSchedulerEndpoint, and YarnSchedulerEndpoint will `askWithReply` KillExecutors to AM. 

Because AM and YarnSchedulerEndpoint are not in the same node in yarn-client mode, it needs to go through the network and wait for a reply.

Since it's a blocking method, I need to call it asynchronously in another thread.